src/D3MHub.sol

```diff
@@ -221,121 +219,125 @@ contract D3MHub {
+            uint256 fixArt = ink - art; // Amount of fees + permissionless DAI paid we will now transform to debt
+            art += fixArt;
+            require(art <= MAXINT256, "D3MHub/overflow");
+            vat.suck(_vow, _vow, fixArt * RAY); // This needs to be done to make sure we can deduct sin[vow] and vice in the next call
+            vat.grab(ilk, address(_pool), address(_pool), _vow, 0, int256(fixArt)); // Generating the debt
         }
 
-        // To save gas you can bring the fees back with the unwind
-        uint256 total = amount + fees;
-        _pool.withdraw(total);
-        daiJoin.join(address(this), total);
+        // Determine if it needs to unwind or wind
 
-        // normalized debt == erc20 DAI to pool (Vat rate for D3M ilks fixed to 1 RAY)
+        uint256 Line;
+        uint256 debt;
+        uint256 targetAssets;
+        uint256 toUnwind;
 
-        if (mode == Mode.NORMAL) {
-            vat.frob(ilk, address(_pool), address(_pool), address(this), -int256(amount), -int256(amount));
-            vat.slip(ilk, address(_pool), -int256(amount));
-            vat.move(address(this), vow, fees * RAY);
-        } else if (mode == Mode.D3M_CULLED) {
-            vat.slip(ilk, address(_pool), -int256(amount));
-            vat.move(address(this), vow, total * RAY);
+        // Determine if it needs to fully unwind due D3M ilk being caged (but not culled) or plan is not active
+        if (ilks[ilk].tic != 0 || !ilks[ilk].plan.active()) {
+            toUnwind = type(uint256).max; // We make sure to enter the unwind path
         } else {
-            // This can be done with the assumption that the price of 1 collateral unit equals 1 DAI.
-            // That way we know that the prev End.skim call kept its gap[ilk] emptied as the CDP was always collateralized.
-            // Otherwise we couldn't just simply take away the collateral from the End module as the next line will be doing.
-            vat.slip(ilk, address(_end), -int256(amount));
-            vat.move(address(this), vow, total * RAY);
-        }
+            Line = vat.Line();
+            debt = vat.debt();
+            targetAssets = ilks[ilk].plan.getTargetAssets(currentAssets);
 
-        emit Unwind(ilk, amount, fees);
-    }
+            // Determine if it needs to unwind due ilk debt ceiling
+            if (art > lineWad) {
+                unchecked {
+                    toUnwind = art - lineWad;
+                }
+            }
+            // Determine if it needs to unwind due global debt ceiling
+            if (debt > Line) {
+                unchecked {
+                    toUnwind = _max(toUnwind, _divup(debt - Line, RAY));
+                }
+            }
+            // Determine if it needs to unwind due plan targetAssets
+            if (targetAssets < currentAssets) {
+                unchecked {
+                    toUnwind = _max(toUnwind, currentAssets - targetAssets);
+                }
+            }
```

Comment:
Consider putting all of this under one `unchecked` block.
Will make the code more compact and visually more similar to the winding logic below.

Reply:
As much as it is a little verbose, I don't like having too large of unchecked blocks as your default mindset is that everything is checked. This exception seems small enough that it's probably okay though.

Reply:
@talbaneth @hexonaut let me know what you think about it now. I feel I got a much better symmetry just with a bit extra of gas usage.

Reply:
I like it, I think you can even make the order of the conditions the same as in wind.

Reply:
This guy Tal is hard to get happy eh! lol joking :)
Change done!

Reply:
lol, perfect now :)